web/component/ad/view.jsx

```diff
@@ -19,11 +19,34 @@ type Props = {
   uri?: string,
   tileLayout?: boolean,
   shouldShowAds: boolean,
+  homepageData?: ?any,
+  claim: any,
 };
 
 function Ad(props: Props) {
-  const { type, uri, tileLayout, shouldShowAds } = props;
+  const { type, uri, tileLayout, shouldShowAds, homepageData, claim } = props;
+  const { categories } = homepageData || {};
   const device = useIsMobile() ? 'mobile' : 'desktop';
+  const channelId =
+    claim && claim.value_type === 'channel' ? claim.claim_id : claim ? claim.signing_channel?.claim_id : undefined;
+
+  const channeldWhitelist = React.useMemo(() => {
+    if (claim && categories) {
+      let channels = [];
+      for (let category in categories) {
+        if (categories[category].channelIds) {
+          for (let channel of categories[category].channelIds) {
+            if (!channels.includes(channel)) {
+              channels.push(channel);
+            }
+          }
+        }
+      }
+      return channels;
+    }
+  }, [categories, claim]);
```

Comment:
This is heavy (over 20ms) and it's also being run 3 times per File Page entry.  The invalidation is due to `claim` being fetched, and `category` being an unstable reference.

But it doesn't need to be re-calculated inside the component.  We can create a cached selector outside like `selectHomepageCategoryChannelIds` that depends on the Db and HomepageCode.  It might still churn a bit at startup, but it won't run 3-times-per-file.

Or I can give it a shot later in the week.